### PR TITLE
Add support for `no.` prefix and omit rangechecks when applicable

### PR DESCRIPTION
**Pull request by johnkellyoxford**
Work in Progress

#17469

### CODE REVIEW

File: src/jit/compiler.hpp [hunk 1]
```diff
@@ -1208,10 +1208,17 @@ inline GenTree* Compiler::gtNewFieldRef(var_types typ, CORINFO_FIELD_HANDLE fldH
...
 {
     GenTreeIndex* gtIndx = new (this, GT_INDEX) GenTreeIndex(typ, arrayOp, indexOp, genTypeSize(typ));
 
+    assert(!(gtIndx->gtFlags & GTF_INX_RNGCHK));
+
+    if (rngChkRequired)
+    {
+        gtIndx->gtFlags |= GTF_INX_RNGCHK;
+    }
+
```
Comment (AndyAyersMS) [inline src/jit/compiler.hpp]: I'd leave this as is (the flag gets set by the CTOR which is fine and normal; you should clear the flag in the importer code after the node is created, when we have no rangecheck prefix.

**Comment by AndyAyersMS** [comment]
You are welcome to pick up my version of your changes: [noprefix](https://github.com/AndyAyersMS/coreclr/tree/noprefix). It gets further but now hits an assert in morph as the "no range check" path in the jit seems to have bit rotted.

I would put up a PR for your fork but since you did your work in master you might want to move it to a branch first (most folks don't do active dev work on the master branches of their forks).

File: src/inc/opcode.def [hunk 2]
```diff
@@ -317,7 +316,7 @@ OPDEF(CEE_INITOBJ,                    "initobj",          PopI,               Pu
 OPDEF(CEE_CONSTRAINED,                "constrained.",     Pop0,               Push0,       InlineType,         IPrefix,     2,  0xFE,    0x16,    META)
 OPDEF(CEE_CPBLK,                      "cpblk",            PopI+PopI+PopI,     Push0,       InlineNone,         IPrimitive,  2,  0xFE,    0x17,    NEXT)
 OPDEF(CEE_INITBLK,                    "initblk",          PopI+PopI+PopI,     Push0,       InlineNone,         IPrimitive,  2,  0xFE,    0x18,    NEXT)
-OPDEF(CEE_UNUSED69,                   "unused",           Pop0,               Push0,       InlineNone,         IPrimitive,  2,  0xFE,    0x19,    NEXT)
+OPDEF(CEE_NO,                         "no.",              Pop0,               Push0,       ShortInlineI,       IPrefix,     2,  0xFE,    0x19,    META)
```
Comment (4creators) [inline src/inc/opcode.def]: After changing this file you should run scripts to generate other files used by System.Private.CoreLib

**Comment by 4creators** [comment]
In general opcode.def is used as source of information to generate other code files used for both managed and native bits of .NET Core. AFAIR one of the scripts is https://github.com/dotnet/coreclr/blob/master/src/inc/OpCodeGen.pl but the are others as well, however, I cannot point to them without digging deeper into code.

File: src/jit/importer.cpp [hunk 3]
```diff
@@ -9939,6 +9951,52 @@ static void impValidateMemoryAccessOpcode(const BYTE* codeAddr, const BYTE* code
...
+        // 0x04: nullcheck (ldfld, stfld, callvirt, ldvirtftn, ldelem.*, stelem.*, ldelema)
+        (!(omittionValues & NO_PREFIX_NULLCHECK
+        || opcode == CEE_LDFLD || opcode == CEE_STFLD || opcode == CEE_CALLVIRT || opcode == CEE_LDVIRTFTN
+        || (opcode >= CEE_LDELEM_I1 && opcode <= CEE_LDELEM_REF)))
+        
+        ))
+    {
+        BADCODE("Invalid opcode for no. prefix");
+    }
+}
```
[the remark this answers is not in the document]
Comment (johnkellyoxford) [reply src/jit/importer.cpp]: There is a missing bracket there causing the issue, which i neglected to push 😬 . Will split into 3 seperate checks :+1:

File: src/jit/gentree.h [hunk 4]
```diff
@@ -4139,6 +4139,7 @@ struct GenTreeIndex : public GenTreeOp
         }
         else
 #endif
+        if (arr->gtFlags & GTF_INX_RNGCHK)
```
[the remark this answers is not in the document]
Comment (johnkellyoxford) [reply src/jit/gentree.h]: would this just be in `impImportBlockCode` after the `INDEX` node is created in the relevant `case`s, or in a different method? not sure where exactly I would do it, most of the stuff I was doing here was trying to copy how `volatile.` prefix works, as that was my best bet to work out what was going on 😄

[the remark this answers is not in the document]
Comment (johnkellyoxford) [reply src/jit/gentree.h]: would this just be in `impImportBlockCode` after the `INDEX` node is created in the relevant `case`s, or in a different method? not sure where exactly I would do it, most of the stuff I was doing here was trying to copy how `volatile.` prefix works, as that was my best bet to work out what was going on 😄

**Pull request closed without merging**